tests/testthat/test-shinytest2-tm_a_pca.R

```diff
@@ -0,0 +1,171 @@
+testthat::test_that("e2e - tm_a_pca: data extract changes output", {
+  skip_if_too_deep(5)
+
+  require(shinytest2)
+
+  data <- within(teal_data(), {
+    require(nestcolor)
+    require(ggplot2)
+
+    USArrests <- USArrests # nolint: object_name.
+  })
+  datanames(data) <- "USArrests"
+
+  app <- TealAppDriver$new(
+    data = data,
+    modules = tm_a_pca(
+      dat = data_extract_spec(
+        dataname = "USArrests",
+        select = select_spec(
+          choices = variable_choices(
+            data = data[["USArrests"]],
+            c("Murder", "Assault", "UrbanPop", "Rape")
+          ),
+          selected = c("Murder", "Assault"),
+          multiple = TRUE
+        )
+      )
+    )
+  )
+
+  # Data selection (adds rows to tables)
+  app$set_module_input("dat-dataset_USArrests_singleextract-select", c("Murder", "Assault"))
+  app$expect_no_validation_error()
+
+  testthat::expect_match(app$get_active_module_output("tbl_eigenvector"), "Assault")
+
+  testthat::expect_failure(
+    testthat::expect_match(app$get_active_module_output("tbl_eigenvector"), "UrbanPop")
+  )
+
+  app$set_module_input("dat-dataset_USArrests_singleextract-select", c("Murder", "UrbanPop"))
+  app$expect_no_validation_error()
+
+  testthat::expect_match(app$get_active_module_output("tbl_eigenvector"), "UrbanPop")
+
+  # Original Coordinate
+  app$set_module_input("plot_type", "Biplot")
+  app$set_module_input("variables", c("Murder"))
+  app$expect_no_validation_error()
+
+  # Color by
+  app$set_module_input("response-dataset_USArrests_singleextract-select", c("Assault"))
+  app$expect_no_validation_error()
+
+  app$set_module_input("response-dataset_USArrests_singleextract-select", c("Murder"))
+  app$expect_validation_error()
```

Comment:
Some of these validation errors failed for me. But, after I included a `app$wait_for_idle(5000)` after them they passed. Looks like we need to modify the app driver with more wait statements.

```
── Failed tests ──────────────────────────────────────────────────────────────────────────────────────────────────
Failure (test-shinytest2-tm_a_pca.R:49:3): e2e - tm_a_pca: data extract changes output
self$get_html(".shiny-output-error-validation") is not NULL

`actual` is a character vector ('<div class="shiny-html-output plot_out_container shiny-bound-output shiny-output-error shiny-output-error-validation shiny-output-error-try-error" id="teal-main_ui-root-principal_component_analysis-module-pca_plot-plot_out_main" width="100%" aria-live="polite">could not find function "vars" \n when evaluating qenv code:\nrot_vars &lt;- pca$rotation[, c("PC1", "PC2")] %&gt;% dplyr::as_tibble(rownames = "label") %&gt;% dplyr::mutate_at(vars(c("PC1", "PC2")), function(x) r * x/sqrt(max(v_scale)))</div>')
`expected` is NULL
No validation error is observed
Backtrace:
    ▆
 1. └─app$expect_no_validation_error() at test-shinytest2-tm_a_pca.R:49:3
 2.   └─testthat::expect_null(...)

Failure (test-shinytest2-tm_a_pca.R:53:3): e2e - tm_a_pca: data extract changes output
self$get_html(".shiny-output-error-validation") is not NULL

`actual` is a character vector ('<div class="shiny-html-output plot_out_container shiny-bound-output shiny-output-error shiny-output-error-validation shiny-output-error-try-error recalculating" id="teal-main_ui-root-principal_component_analysis-module-pca_plot-plot_out_main" width="100%" aria-live="polite">could not find function "vars" \n when evaluating qenv code:\nrot_vars &lt;- pca$rotation[, c("PC1", "PC2")] %&gt;% dplyr::as_tibble(rownames = "label") %&gt;% dplyr::mutate_at(vars(c("PC1", "PC2")), function(x) r * x/sqrt(max(v_scale)))</div>')
`expected` is NULL
No validation error is observed
Backtrace:
    ▆
 1. └─app$expect_no_validation_error() at test-shinytest2-tm_a_pca.R:53:3
 2.   └─testthat::expect_null(...)

Failure (test-shinytest2-tm_a_pca.R:108:3): e2e - tm_a_pca: encodings update main panel
self$get_html(".shiny-output-error-validation") is not NULL

`actual` is a character vector ('<div class="shiny-html-output plot_out_container shiny-bound-output shiny-output-error shiny-output-error-validation shiny-output-error-try-error" id="teal-main_ui-root-principal_component_analysis-module-pca_plot-plot_out_main" width="100%" aria-live="polite">could not find function "vars" \n when evaluating qenv code:\nrot_vars &lt;- pca$rotation[, c("PC1", "PC2")] %&gt;% dplyr::as_tibble(rownames = "label") %&gt;% dplyr::mutate_at(vars(c("PC1", "PC2")), function(x) r * x/sqrt(max(v_scale)))</div>')
`expected` is NULL
No validation error is observed
Backtrace:
    ▆
 1. └─app$expect_no_validation_error() at test-shinytest2-tm_a_pca.R:108:3
 2.   └─testthat::expect_null(...)

Failure (test-shinytest2-tm_a_pca.R:111:3): e2e - tm_a_pca: encodings update main panel
self$get_html(".shiny-output-error-validation") is not NULL

`actual` is a character vector ('<div class="shiny-plot-output html-fill-item shiny-output-error shiny-output-error-validation shiny-output-error-try-error shiny-bound-output" id="teal-main_ui-root-principal_component_analysis-module-pca_plot-plot_main" style="width:100%;height:100%;" aria-live="polite">could not find function "vars" \n when evaluating qenv code:\nrot_vars &lt;- pca$rotation[, c("PC1", "PC2")] %&gt;% dplyr::as_tibble(rownames = "label") %&gt;% dplyr::mutate_at(vars(c("PC1", "PC2")), function(x) r * x/sqrt(max(v_scale)))</div>')
`expected` is NULL
No validation error is observed
Backtrace:
    ▆
 1. └─app$expect_no_validation_error() at test-shinytest2-tm_a_pca.R:111:3
 2.   └─testthat::expect_null(...)

Failure (test-shinytest2-tm_a_pca.R:168:3): e2e - tm_a_pca: encodings update main panel
self$get_html(".shiny-output-error-validation") is not NULL

`actual` is a character vector ('<div class="shiny-html-output plot_out_container shiny-bound-output shiny-output-error shiny-output-error-validation shiny-output-error-try-error recalculating" id="teal-main_ui-root-principal_component_analysis-module-pca_plot-plot_out_main" width="100%" aria-live="polite">could not find function "vars" \n when evaluating qenv code:\nrot_vars &lt;- pca$rotation[, c("PC1", "PC2")] %&gt;% dplyr::as_tibble(rownames = "label") %&gt;% dplyr::mutate_at(vars(c("PC1", "PC2")), function(x) r * x/sqrt(max(v_scale)))</div>')
`expected` is NULL
No validation error is observed
Backtrace:
    ▆
 1. └─app$expect_no_validation_error() at test-shinytest2-tm_a_pca.R:168:3
 2.   └─testthat::expect_null(...)

[ FAIL 5 | WARN 0 | SKIP 0 | PASS 93 ]

r$>

r$> devtools::test()
ℹ Testing teal.modules.general
✔ | F W  S  OK | Context
✔ |         31 | shinytest2-tm_a_pca [165.8s]
✔ |         18 | tm_g_bivariate
✔ |         21 | bivariate_ggplot_call
✔ |          4 | front_page
✔ |         15 | g_expert_cl
✔ |          6 | scatterplotmatrix_get_stats
✔ |          3 | variable_browser

══ Results ═══════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 166.0 s

[ FAIL 0 | WARN 0 | SKIP 0 | PASS 98 ]
```

Reply:
This is being addressed by adding `app_driver$wait_for_idle()` but can be removed in the future after https://github.com/insightsengineering/teal/pull/1200 is merged

Reply:
Can you check if this still happens?

Reply:
It works fine now. Resolving.

Reply:
This happens for me at https://github.com/insightsengineering/teal.modules.general/pull/722#issuecomment-2058952357 `tm_g_association` even though https://github.com/insightsengineering/teal/pull/1200 is merged

Reply:
Added `app$wait_for_idle()` in 3 places